pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,136 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	"github.com/aws/karpenter/pkg/providers/instance"
+	"github.com/aws/karpenter/pkg/utils"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+const (
+	TagNodeClaim = corev1beta1.Group + "/nodeclaim"
+	TagName      = "Name"
+)
+
+type Controller struct {
+	kubeClient       client.Client
+	instanceProvider *instance.Provider
+}
+
+func NewController(kubeClient client.Client, instanceProvider *instance.Provider) corecontroller.Controller {
+	return corecontroller.Typed[*corev1beta1.NodeClaim](kubeClient, &Controller{
+		kubeClient:       kubeClient,
+		instanceProvider: instanceProvider,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "nodeclaim.tagging"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClaim *corev1beta1.NodeClaim) (reconcile.Result, error) {
+	storedNodeClaim := nodeClaim.DeepCopy()
+
+	if !isTaggable(nodeClaim) {
+		return reconcile.Result{}, nil
+	}
+
+	id, err := utils.ParseInstanceID(nodeClaim.Status.ProviderID)
+	if err != nil {
+		// We don't throw an error here since we don't want to retry until the ProviderID has been updated.
+		logging.FromContext(ctx).Errorf("failed to parse instance ID, %w", err)
+		return reconcile.Result{}, nil
+	}
+
+	err = c.tagInstance(ctx, nodeClaim, id)
+	if err != nil {
+		if !cloudprovider.IsNodeClaimNotFoundError(err) {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	nodeClaim.Annotations = lo.Assign(nodeClaim.Annotations, map[string]string{v1beta1.AnnotationInstanceTagged: "true"})
+	if !equality.Semantic.DeepEqual(nodeClaim, storedNodeClaim) {
+		if err := c.kubeClient.Patch(ctx, nodeClaim, client.MergeFrom(storedNodeClaim)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				nc := o.(*corev1beta1.NodeClaim)
+				return isTaggable(nc)
+			})),
+	)
+}
+
+func (c *Controller) tagInstance(ctx context.Context, nc *corev1beta1.NodeClaim, id string) error {
+	tags := map[string]string{
+		TagName:      nc.Status.NodeName,
+		TagNodeClaim: nc.Name,
+	}
+
+	// Remove tags which have been already populated
+	instance, err := c.instanceProvider.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+	tags = lo.OmitByKeys(tags, lo.Keys(instance.Tags))
+
+	if len(tags) != 0 {
+		err = c.instanceProvider.CreateTags(ctx, id, tags)
```

Comment:
Creating tags should be idempotent, which means we shouldn't need to do another instanceProvider.Get() call to get the tags here. Can we just do the `createTags` call here without a `instanceProvider.Get()` since we're already doing a short circuit in the controller with the node annotation?

Reply:
We don't want to overwrite tags that the user specifies. A use case scenario for this would be if a user specifies a `Name` tag as part of the `NodeClass` or if they have a separate process that applies the tag. Unless we want to say that Karpenter owns the `Name` tag and it can't be configured by users we need some mechanism to determine if we should tag it. In this case, we check if it has already been tagged.

Reply:
Could we just check if the user has defined any tags we don't want to override in their NodeClass and then just remove that from our CreateTags call, rather than doing Get() and then Create()? Trying to compare it to the underlying instance means we have to sync and union the tags each time, when I think we can save on API calls by just using the information that Karpenter already knows and not adding that to the createTags call in the first place.

Reply:
The case that doesn't necessarily cover is if an external process is responsible for tagging. In that case, maybe users should specify a 'default' tag as part of the NodeClass that way we know not to tag it? That seems like a reasonable solution to me, but maybe a little messy?

Reply:
What if we just make sure that we're only adding tags with a Karpenter prefix? That way they can add the Name tag if they want to, but it'll be a different name tag than the one that we set (maybe something like `karpenter.sh/node-name`)

Reply:
I think that would be reasonable. However, if in the future we want to support additional dynamic tags would it make sense for those to live in the `karpenter.sh` namespace? Also, it would still be nice to support the `Name` tag since that's the visible name for the instance in the AWS console. Maybe we could special case ownership for `Name`? Thoughts @jonathan-innis?

Reply:
Alligned offline to keep original implementation (not overwriting existing tags). This model ensures that user specified tags are not overwritten while also giving us the flexibility to create tags not within the karpenter namespace (e.g. `Name`).